### PR TITLE
Fix inliners to run all passes on IR and clean up correctly.

### LINKED ISSUE #4691: tuple slicing fails in inlined functions
Before submitting a bug report please ensure that you can check off these boxes:

- [x] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).

0.46.0

- [x] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

Here is the reproducer:

```python
import numpy as np
from numba import njit, types
from numba.extending import overload


def output_factory(array, dtype):
    pass


@overload(output_factory, inline='always')
def ol_output_factory(array, dtype):
    if isinstance(array, types.npytypes.Array):
        def impl(array, dtype):
            shape = array.shape[3:]
            return np.zeros(shape, dtype=dtype)

        return impl


@njit(nogil=True)
def fn(array):
    out = output_factory(array, array.dtype)
    return out


@njit(nogil=True)
def fn2(array):
    return np.zeros(array.shape[3:], dtype=array.dtype)


fn(np.ones((10, 20, 30, 40, 50)))

fn2(np.ones((10, 20, 30, 40, 50)))
```

The call to ``fn`` fails with the following:

```traceback
Traceback (most recent call last):
  File "test_inlining.py", line 30, in <module>
    fn(np.ones((10, 20, 30, 40, 50)))
  File "/home/sperkins/venv/afr3/lib/python3.6/site-packages/numba/dispatcher.py", line 401, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/home/sperkins/venv/afr3/lib/python3.6/site-packages/numba/dispatcher.py", line 344, in error_rewrite
    reraise(type(e), e, None)
  File "/home/sperkins/venv/afr3/lib/python3.6/site-packages/numba/six.py", line 668, in reraise
    raise value.with_traceback(tb)
numba.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of Function(<function output_factory at 0x7f9f5ac4ce18>) with argument(s) of type(s): (array(float64, 5d, C), dtype(float64))
 * parameterized
In definition 0:
    TypingError: Invalid use of Function(<built-in function getitem>) with argument(s) of type(s): (tuple(int64 x 5), slice<a:b>)
 * parameterized
In definition 0:
    All templates rejected with literals.
In definition 1:
    All templates rejected without literals.
In definition 2:
    All templates rejected with literals.
In definition 3:
    All templates rejected without literals.
In definition 4:
    All templates rejected with literals.
In definition 5:
    All templates rejected without literals.
In definition 6:
    All templates rejected with literals.
In definition 7:
    All templates rejected without literals.
In definition 8:
    All templates rejected with literals.
In definition 9:
    All templates rejected without literals.
In definition 10:
    All templates rejected with literals.
In definition 11:
    All templates rejected without literals.
In definition 12:
    All templates rejected with literals.
In definition 13:
    All templates rejected without literals.
This error is usually caused by passing an argument of a type that is unsupported by the named function.
[1] During: typing of intrinsic-call at test_inlining.py (14)

File "test_inlining.py", line 14:
        def impl(array, dtype):
            shape = array.shape[3:]
            ^

    raised from /home/sperkins/venv/afr3/lib/python3.6/site-packages/numba/typeinfer.py:951
In definition 1:
    TypingError: Invalid use of Function(<built-in function getitem>) with argument(s) of type(s): (tuple(int64 x 5), slice<a:b>)
 * parameterized
In definition 0:
    All templates rejected with literals.
In definition 1:
    All templates rejected without literals.
In definition 2:
    All templates rejected with literals.
In definition 3:
    All templates rejected without literals.
In definition 4:
    All templates rejected with literals.
In definition 5:
    All templates rejected without literals.
In definition 6:
    All templates rejected with literals.
In definition 7:
    All templates rejected without literals.
In definition 8:
    All templates rejected with literals.
In definition 9:
    All templates rejected without literals.
In definition 10:
    All templates rejected with literals.
In definition 11:
    All templates rejected without literals.
In definition 12:
    All templates rejected with literals.
In definition 13:
    All templates rejected without literals.
This error is usually caused by passing an argument of a type that is unsupported by the named function.
[1] During: typing of intrinsic-call at test_inlining.py (14)

File "test_inlining.py", line 14:
        def impl(array, dtype):
            shape = array.shape[3:]
            ^

    raised from /home/sperkins/venv/afr3/lib/python3.6/site-packages/numba/typeinfer.py:951
This error is usually caused by passing an argument of a type that is unsupported by the named function.
[1] During: resolving callee type: Function(<function output_factory at 0x7f9f5ac4ce18>)
[2] During: typing of call at test_inlining.py (22)


File "test_inlining.py", line 22:
def fn(array):
    out = output_factory(array, array.dtype)
    ^

This is not usually a problem with Numba itself but instead often caused by
the use of unsupported features or an issue in resolving types.

To see Python/NumPy features supported by the latest release of Numba visit:
http://numba.pydata.org/numba-doc/latest/reference/pysupported.html
and
http://numba.pydata.org/numba-doc/latest/reference/numpysupported.html

For more information about typing errors and how to debug them visit:
http://numba.pydata.org/numba-doc/latest/user/troubleshoot.html#my-code-doesn-t-compile

If you think your code should work with Numba, please report the error message
and traceback, along with a minimal reproducer at:
https://github.com/numba/numba/issues/new
```

but the call to ``fn2`` succeeds.

I know the ``inline`` feature is new. Am I perhaps using it incorrectly?

### LINKED ISSUE #4693: raising Exceptions in inlined functions fails on compile
- [x] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).

0.46.0

- [x] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

The below:

```python

from numba import njit, generated_jit
from numba import types
import numpy as np


@njit(inline='always')
def inlining(array):
    if array.ndim != 1:
        raise ValueError("Invalid number of dimensions")

    return array

@njit
def fn(array):
    return inlining(array)

print(fn(np.zeros(10)))
```

fails with

```traceback
Traceback (most recent call last):
  File "test_inlining_generated_jit.py", line 19, in <module>
    print(fn(np.zeros(10)))
  File "/home/sperkins/venv/afr3/lib/python3.6/site-packages/numba/dispatcher.py", line 404, in _compile_for_args
    error_rewrite(e, 'unsupported_error')
  File "/home/sperkins/venv/afr3/lib/python3.6/site-packages/numba/dispatcher.py", line 344, in error_rewrite
    reraise(type(e), e, None)
  File "/home/sperkins/venv/afr3/lib/python3.6/site-packages/numba/six.py", line 668, in reraise
    raise value.with_traceback(tb)
numba.errors.UnsupportedError: Failed in nopython mode pipeline (step: nopython frontend)
Unsupported constraint encountered: raise $10.3.7

File "test_inlining_generated_jit.py", line 9:
def inlining(array):
    <source elided>
    if array.ndim != 1:
        raise ValueError("Invalid number of dimensions")
        ^

Unsupported functionality was found in the code Numba was trying to compile.

If this functionality is important to you please file a feature request at:
https://github.com/numba/numba/issues/new
```

but succeeds when `inline='never'`

### LINKED ISSUE #5476: Raising of an exception in inlined function fails on compilation
- [ ] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [X] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

```python
from numba import njit

@njit(inline='always')
def inlining(msg):
    raise ValueError(msg)

@njit
def fn():
    return inlining('Something happened')

print(fn())
```

Traceback:
```
Traceback (most recent call last):
  File "test_inline.py", line 14, in <module>
    print(fn())
  File "numba\numba\dispatcher.py", line 420, in _compile_for_args
    raise e
  File "numba\numba\dispatcher.py", line 353, in _compile_for_args
    return self.compile(tuple(argtypes))
  File "numba\numba\compiler_lock.py", line 32, in _acquire_compile_lock
    return func(*args, **kwargs)
  File "numba\numba\dispatcher.py", line 768, in compile
    cres = self._compiler.compile(args, return_type)
  File "numba\numba\dispatcher.py", line 77, in compile
    status, retval = self._compile_cached(args, return_type)
  File "numba\numba\dispatcher.py", line 91, in _compile_cached
    retval = self._compile_core(args, return_type)
  File "numba\numba\dispatcher.py", line 109, in _compile_core
    pipeline_class=self.pipeline_class)
  File "numba\numba\compiler.py", line 551, in compile_extra
    return pipeline.compile_extra(func)
  File "numba\numba\compiler.py", line 331, in compile_extra
    return self._compile_bytecode()
  File "numba\numba\compiler.py", line 393, in _compile_bytecode
    return self._compile_core()
  File "numba\numba\compiler.py", line 373, in _compile_core
    raise e
  File "numba\numba\compiler.py", line 364, in _compile_core
    pm.run(self.state)
  File "numba\numba\compiler_machinery.py", line 347, in run
    raise patched_exception
  File "numba\numba\compiler_machinery.py", line 338, in run
    self._runPass(idx, pass_inst, state)
  File "numba\numba\compiler_lock.py", line 32, in _acquire_compile_lock
    return func(*args, **kwargs)
  File "numba\numba\compiler_machinery.py", line 302, in _runPass
    mutated |= check(pss.run_pass, internal_state)
  File "numba\numba\compiler_machinery.py", line 275, in check
    mangled = func(compiler_state)
  File "numba\numba\untyped_passes.py", line 312, in run_pass
    state.func_ir.blocks = simplify_CFG(state.func_ir.blocks)
  File "numba\numba\ir_utils.py", line 1237, in simplify_CFG
    return rename_labels(blocks)
  File "numba\numba\ir_utils.py", line 1188, in rename_labels
    topo_order.remove(return_label)
ValueError: Failed in nopython mode pipeline (step: inline inlinable functions)
list.remove(x): x not in list
```

 I faced the issue accidentally and unfortunately don't know real use case of that.
Numba version: ef119bcd1733ff49d71bdf2da8a66e91bb704f83

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title.

Fixes #4691
Fixes #4693
Fixes #5476

### CODE REVIEW

File: numba/core/typing/templates.py [hunk 1]
```diff
@@ -527,9 +544,9 @@ def generic(self, args, kws):
                 self._compiled_overloads[sig.args] = disp_type.get_overload(sig)
                 # store the inliner information, it's used later in the cost
                 # model function call
-                iinfo = _inline_info(ir, typemap, calltypes, sig)
-                self._inline_overloads[sig.args] = {'folded_args': folded_args,
-                                                    'iinfo': iinfo}
+            iinfo = _inline_info(ir, typemap, calltypes, sig)
+            self._inline_overloads[sig.args] = {'folded_args': folded_args,
```
Comment (stuartarchibald) [inline numba/core/typing/templates.py:548]: this overrides the behaviour here https://github.com/numba/numba/pull/5673/files#diff-aee5ebdb5580a36762cba728715f96d1R536-R537 and needs looking at

File: numba/core/inline_closurecall.py [hunk 2]
```diff
@@ -262,6 +269,302 @@ def check_reduce_func(func_ir, func_var):
...
+        state.func_ir = None
+        state.typingctx = self.typingctx
+        state.targetctx = self.targetctx
+        state.locals = self.locals
+        state.pipeline = self.pipeline
+        state.flags = self.flags
+
+        # Disable SSA transformation, the call site won't be in SSA form and
+        # self.inline_ir depends on this being the case.
+        state.flags.enable_ssa = False
```
Comment (stuartarchibald) [inline numba/core/inline_closurecall.py:501]: Is this true for overload inlines?

File: numba/core/typing/templates.py [hunk 3]
```diff
@@ -500,7 +500,24 @@ def generic(self, args, kws):
...
             from numba.core import typed_passes, compiler
-            ir = compiler.run_frontend(disp_type.dispatcher.py_func)
+            from numba.core.inline_closurecall import InlineWorker
+            fcomp = disp._compiler
+            flags = compiler.Flags()
+
+            # Updating these causes problems?!
+            #fcomp.targetdescr.options.parse_as_flags(flags,
+            #                                         fcomp.targetoptions)
+            #flags = fcomp._customize_flags(flags)
```
Comment (stuartarchibald) [inline numba/core/typing/templates.py:510]: Is this a problem? Do flags influence IR gen?

**Comment by stuartarchibald** [comment]
https://github.com/stuartarchibald/numba/commit/94f6bb27b7e336a5efc8efde3a126e49652ca7ad achieves the same/similar, but as a strategy pushes inlining as high up the compilation chain as possible, and then just lets other passes run as usual. The inlined IR is not subjected to the full compilation pipelines, only closure inlining and generic rewrites.

File: numba/core/inline_closurecall.py (hunk 2 above)
Comment (sklam) [reply numba/core/inline_closurecall.py:501]: why is this needed?

File: numba/core/typing/templates.py (hunk 3 above)
Comment (sklam) [reply numba/core/typing/templates.py:510]: only in `cfuncs` it seems

File: numba/core/inline_closurecall.py [hunk 4]
```diff
@@ -262,6 +269,302 @@ def check_reduce_func(func_ir, func_var):
...
+                   "calltypes missing in initialization.")
+            raise ValueError(msg)
+        from numba.core import typed_passes
+        # call branch pruning to simplify IR and avoid inference errors
+        callee_ir._definitions = ir_utils.build_definitions(callee_ir.blocks)
+        numba.core.analysis.dead_branch_prune(callee_ir, arg_typs)
+        try:
+            f_typemap, f_return_type, f_calltypes = typed_passes.type_inference_stage(
+                    self.typingctx, callee_ir, arg_typs, None)
+        except Exception:
```
Comment (sklam) [inline numba/core/inline_closurecall.py:538]: pretty certain that you don't need this try..except. It's never exercised by the testsuite.

**Comment by sklam** [comment]
Note: failing at https://github.com/numba/numba/blob/ff24fb1d32e83602742b548ad1c3aa240f8b34b4/numba/tests/test_mixed_tuple_unroller.py#L1586-L1599 because of `literal_unroll(x)` where `x = [10, y]`

**Comment by sklam** [comment]
NOTE: prefer this PR over https://github.com/stuartarchibald/numba/commit/94f6bb27b7e336a5efc8efde3a126e49652ca7ad

File: numba/core/inline_closurecall.py (hunk 2 above)
Comment (stuartarchibald) [reply numba/core/inline_closurecall.py:501]: without it, this:
```
======================================================================
ERROR: test_inlining_optional_constant (numba.tests.test_ir_inlining.TestGeneralInlining)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "<path>numba/tests/test_ir_inlining.py", line 1043, in test_inlining_optional_constant
    self.check(impl, block_count='SKIP', inline_expect={'bar': True})
  File "<path>numba/tests/test_ir_inlining.py", line 103, in check
    self.assertEqual(test_impl(*args), j_func(*args))
  File "<path>numba/core/dispatcher.py", line 401, in _compile_for_args
    error_rewrite(e, 'typing')
  File "<path>numba/core/dispatcher.py", line 344, in error_rewrite
    reraise(type(e), e, None)
  File "<path>numba/core/utils.py", line 79, in reraise
    raise value.with_traceback(tb)
numba.core.errors.TypingError: Failed in inliner_custom_pipe mode pipeline (step: nopython frontend)
Type of variable 'b.1.382' cannot be determined, operation: unknown operation, location: unknown location (0:0)

File "unknown location", line 0:
<source missing, REPL/exec in use?>
```

File: numba/core/inline_closurecall.py (hunk 4 above)
Comment (stuartarchibald) [reply numba/core/inline_closurecall.py:538]: I think you are right, I've a feeling this is an artifact of development from the numerous rounds of trying to get typed inlining working.

File: numba/core/typing/templates.py (hunk 3 above)
Comment (stuartarchibald) [reply numba/core/typing/templates.py:510]: leave it for now?

**Comment by stuartarchibald** [comment]
> test_inlined_unroll_list

elect to skip for now, needs updates elsewhere

**Comment by sklam** [comment]
is this ready for review?

**Comment by stuartarchibald** [comment]
> is this ready for review?

Let's say yes to get it moving, but also note there's a couple of thing we may need to look at already mentioned in the above.

File: numba/core/inline_closurecall.py [hunk 5]
```diff
@@ -262,6 +269,297 @@ def check_reduce_func(func_ir, func_var):
...
+        `block`. If `arg_typs` is given and the InlineWorker instance was
+        initialized with a typemap and calltypes then they will be appropriately
+        updated based on the arg_typs.
+        """
+        callee_ir = self.run_untyped_passes(function)
+        freevars = function.__code__.co_freevars
+        return self.inline_ir(caller_ir, block, i, callee_ir, freevars,
+                              arg_typs=arg_typs)
+
+    def inline_closure(self, caller_ir, block, i, closure_obj, glbls,
```
Comment (sklam) [inline numba/core/inline_closurecall.py:448]: It seems like this function is dead

File: numba/core/inline_closurecall.py [hunk 6]
```diff
@@ -262,6 +269,297 @@ def check_reduce_func(func_ir, func_var):
...
+        canonicalize_array_math(callee_ir, f_typemap,
+                                f_calltypes, self.typingctx)
+        # remove argument entries like arg.a from typemap
+        arg_names = [vname for vname in f_typemap if vname.startswith("arg.")]
+        for a in arg_names:
+            f_typemap.pop(a)
+        self.typemap.update(f_typemap)
+        self.calltypes.update(f_calltypes)
+
+    def convert_code_object_to_function(self, obj, glbls):
```
Comment (sklam) [inline numba/core/inline_closurecall.py:546]: this too

**Comment by stuartarchibald** [comment]
I added some tests for the dead functions to the  tip of this branch: https://github.com/stuartarchibald/numba/commits/fix/5476_5_cont

My conclusion from adding them is that the interface still isn't quite right and that it'd make sense to also add to the class something that actually does the inlining work (stitches the blocks together etc), as well as computing the changes required. I'm going to remove those dead methods from the class as a result and open a ticket about future things to be done.

**Comment by stuartarchibald** [comment]
Work ticket: https://github.com/numba/numba/issues/5775

**Pull request merged**